src/core-recipe.jl

```diff
@@ -62,17 +62,37 @@ macro plot_or_defaults(var, defaults)
     return :(plot_or_defaults($(esc(var)), $(esc(defaults)), $(QuoteNode(var))))
 end
 
+get_bounding_rect(rect::Rect) = rect
+
+function get_bounding_rect(circ::Circle)
+    xmin, ymin = minimum(circ)
+    xmax, ymax = maximum(circ)
+    Rect2f(xmin, ymin, xmax - xmin, ymax - ymin)
+end
+
+function get_bounded(rect, positions, data, padding, pad_value)
+    bb_xy = decompose(Point2f, rect)
+    bb_v = fill(pad_value, 4)
+    @assert padding ≥ 0 "cannot deal with negative padding"
+    @assert allunique(positions) "positions must be unique"
```

Comment:
```suggestion
    padding ≥ 0 || error("cannot deal with negative padding")
    allunique(positions) || error("positions must be unique")
```

`@assert` can be optimized out in some situations